### PR TITLE
[Security] Deprecate "AbstractVoter" in favor of "Voter"

### PR DESCRIPTION
**Pull request by lyrixx**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #16556, #16558, #16554 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php [hunk 1]
```diff
@@ -53,10 +53,10 @@ public function supportsClass($class);
      * ACCESS_GRANTED, ACCESS_DENIED, or ACCESS_ABSTAIN.
      *
      * @param TokenInterface $token      A TokenInterface instance
-     * @param object|null    $object     The object to secure
+     * @param mixed          $subject    The subject to secure
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php]: this is a BC break, as explained in #16600

**Comment by stof** [comment]
I don't see any reason to deprecated AbstractVoter in favor of an abstract Voter class (which does not follow our naming conventions btw). This does not remove the BC break

**Comment by lyrixx** [comment]
> I don't see any reason to deprecated AbstractVoter

I deprecated it because I added a new "base voter". 
I added a new "base voter" in order to fix listed issue.

**Comment by lyrixx** [comment]
I reverted the change on the interface to ease integration of this PR.

**Comment by nicolas-grekas** [comment]
:+1:
Status: reviewed

It removes a limitation of the current AbstractVoter (the first `if` block of the `vote` method) that can't be fixed with a simple deprecation notice. Which means we need a new class name. `Voter` is fine to me (abstract, like `Command` is).
@weaverryan I'd be happy to have your vote here since you cared about the linked issue.

File: UPGRADE-2.8.md [hunk 2]
```diff
@@ -475,6 +475,8 @@ Security
    }
    ```
 
+ * The `AbstractVoter` is deprecated. Upgrade to `Voter` instead.
+
```
Comment (weaverryan) [inline UPGRADE-2.8.md]: If we're deprecating `AbstractVoter` entirely, then we don't need to talk about it at all, right? The upgrade path would be to switch to `Voter` and make the changes needed there, correct? If so, I think we should have one simple, before and after (before `AbstractVoter` after with `Voter`).

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ */
+abstract class Voter implements VoterInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsAttribute($attribute)
+    {
+        throw new \BadMethodCallException('supportsAttribute method is deprecated since version 2.8, to be removed in 3.0');
+    }
```
Comment (weaverryan) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: If the `AbstractVoter` still works in 2.8 (but is deprecated of course), do we need these methods at all in the new class? When I switch to the new class, I will change to use the new methods.

**Comment by weaverryan** [comment]
I left a few comments, but I'm "pro" this idea and I like the name.

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php (hunk 3 above)
Comment (lyrixx) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: Yes, to **implements** the interface :/

File: UPGRADE-2.8.md (hunk 2 above)
Comment (lyrixx) [inline UPGRADE-2.8.md]: If people were using AbstractVoter in 2.7, they can read how to upgrade to 2.8;
Then they could upgrade with the new Voter. Like that it's done step by step.

But I can merge all steps together.

Comment (nicolas-grekas) [inline UPGRADE-2.8.md]: Or, we could and I think niw should juste revert the changes made on AbstractVoter for 2.8 entirely, but keep the deprecations. Ie remove the supports method & co. Don't you think ?

Comment (lyrixx) [inline UPGRADE-2.8.md]: This is a  real BC break

Comment (nicolas-grekas) [inline UPGRADE-2.8.md]: How that? SF 2.8 is not released, how can it be à BC break to revert the changes made on 2.8?

Comment (lyrixx) [inline UPGRADE-2.8.md]: Because the implementation sticks on the implementation of the 2.7 version.

**Comment by weaverryan** [comment]
But I also think this can wait until 3.0... It's great, but I'm not sure it's important enough to stick into the release at the last second: there's too much else going on. If we do save this for 3.1, then I can update the `AbstractVoter` docs (I can't work on that til this gets a decision).

**Comment by weaverryan** [comment]
Whoops - I meant wait until 3.1 - I updated my comment.

**Comment by nicolas-grekas** [comment]
We worked together with @lyrixx on this, the PR now has two commits:
- the first one reverts all changes made between 2.7 and 2.8-beta on AbstractVoter and deprecates it instead;
- the second one moves all this work made between 2.7 and 2.8-beta on the new `Voter` class.

The new `Voter` class is now more generic, at least it is generic enough to be reused by all the concrete voter we provide by default (to be done in an other PR).

We can postpone this deprecation to 3.1, but it would be quite unfortunate to ask devs to update their app twice. It's late but maybe not too late to fix this limitation now. That would reduce the burden significantly on this topic.


```diff
git diff symfony/2.7...d3c6d93dff9761cacc811976656574985fb37009 -- src/Symfony/Component/Security/Core/Authorization/Voter/
diff --git a/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php b/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
index efa1562..5dcf787 100644
--- a/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
+++ b/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
@@ -11,6 +11,8 @@
 
 namespace Symfony\Component\Security\Core\Authorization\Voter;
 
+@trigger_error('The '.__NAMESPACE__.'\AbstractVoter class is deprecated since version 2.8, to be removed in 3.0. Upgrade to Symfony\Component\Security\Core\Authorization\Voter\Voter instead.', E_USER_DEPRECATED);
+
 use Symfony\Component\Security\Core\User\UserInterface;
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
 
@@ -18,6 +20,8 @@ use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
  * Abstract Voter implementation that reduces boilerplate code required to create a custom Voter.
  *
  * @author Roman Marint<C5><A1>enko <inoryy@gmail.com>
+ *
+ * @deprecated since version 2.8, to be removed in 3.0. Upgrade to Symfony\Component\Security\Core\Authorization\Voter\Voter instead.
  */
 abstract class AbstractVoter implements VoterInterface
 {
diff --git a/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php b/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
index d00ff1c..7e243f9 100644
--- a/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
+++ b/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
@@ -30,6 +30,8 @@ interface VoterInterface
      * @param string $attribute An attribute
      *
      * @return bool true if this Voter supports the attribute, false otherwise
+     *
+     * @deprecated since version 2.8, to be removed in 3.0.
      */
     public function supportsAttribute($attribute);
 
@@ -39,6 +41,8 @@ interface VoterInterface
      * @param string $class A class name
      *
      * @return bool true if this Voter can process the class
+     *
+     * @deprecated since version 2.8, to be removed in 3.0.
      */
     public function supportsClass($class);
 ```

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        foreach ($attributes as $attribute) {
+            if (!$this->supports($attribute, $object)) {
+                continue;
+            }
+
+            // as soon as at least one attribute is supported, default is to deny access
+            $vote = self::ACCESS_DENIED;
+
+            if ($this->voteOnAttribute($attribute, $object, $token)) {
+                // grant access as soon as at least one voter returns a positive response
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: this should be ``at least one attribute``. We are inside the voter here. We don't have multiple voters

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                return self::ACCESS_GRANTED;
+            }
+        }
+
+        return $vote;
+    }
+
+    /**
+     * Determines if the attribute and object are supported by this voter.
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: ``whether``, not ``if`` AFAICT

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param string $attribute An attribute
+     * @param mixed  $object    The object to secure
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: this should be renamed if it is not always an object

Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: Well, grammatically, it's an object. But a synonym would be less confusing I agree. [Here](http://dico.isc.cnrs.fr/dico/tr/search_en?r=object&send=Translate) are some, any preference?

Comment (jvasseur) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: I think subject could be used instead.

Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: I just updated the PR to use "matter" instead, [as suggested by thesaurus.com(http://www.thesaurus.com/browse/object).

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: fixed

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: I kept `if`. I'd use whether if it were "attribute **or** object", but with "and", it sounds better to me. I a native wants to balance this, I'd be happy to have their feedback

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php (hunk 6 above)
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: `$matter` is not a common term and it may be difficult to understand it. Could we call it `$object` and add a note in the PHPdocs saying that this can also be a non-object? By the way, how common is to pass a non-object to this argument?

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * Determines if the attribute and matter are supported by this voter.
+     *
+     * @param string $attribute An attribute
+     * @param mixed  $matter    The matter to secure
+     *
+     * @return bool True if the attribute and matter are supported, false otherwise
+     */
+    abstract protected function supports($attribute, $matter);
```
Comment (weaverryan) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: `$matter` is a weird word in English - usually it only refers to the scientific meaning of matter (https://en.wikipedia.org/wiki/Matter).

I liked `$subject` better, and I think we should hint in the docs that this is often an object:

```php
/**
 * @param mixed $subject   The subject to secure (e.g. an object the user wants to access)
 */
```

Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: Lets go for subject, thanks for the feedback. I used `The subject to secure, e.g. an object the user wants to access or any other PHP type`

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php (hunk 6 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php]: @javiereguiluz frequent enough to raise all this discussion and the related issues :)

**Comment by nicolas-grekas** [comment]
Thank you all for the feedback, all comments are now resolved, votes pending :)

**Comment by fabpot** [comment]
So, IIUC, the `AbstractVoter` class is the same as in 2.7 with the deprecation notices added and the new `Voter` class only removed the `if` statement that was problematic. Looks like a no-brainer to me. :+1:

**Comment by weaverryan** [comment]
:+1: I agree - that bug was always something we wanted to fix - we should have just done it all at once (like this PR).

**Comment by nicolas-grekas** [comment]
Thank you @lyrixx.
